docs/standard/microservices-architecture/multi-container-microservice-net-applications/implementing-api-gateways-with-ocelot.md

```diff
@@ -0,0 +1,579 @@
+
+Going much further in the design, sometimes a fine-grained API Gateway can also be limited to a single business microservice depending on the chosen architecture. Having the API Gateway’s boundaries dictated by the business or domain will help you to get a better design. 
+
+For instance, fine granularity in the API Gateway tier can be especially useful for more advanced composite UI applications that are based on microservices, because the concept of a fine-grained API Gateway is similar to a UI composition service. 
+
+You can read further information about this topic in the section [Creating composite UI based on microservices](https://docs.microsoft.com/dotnet/standard/microservices-architecture/architect-microservice-container-applications/microservice-based-composite-ui-shape-layout).
+
+As key takeaway, for many medium- and large-size applications, using a custom-built API Gateway product is usually a good approach, but not as a single monolithic aggregator or unique central custom API Gateway unless that API Gateway allows multiple independent configuration areas for the several development teams creating autonomous microservices.
+
+
+### Sample microservices/containers to reroute through the API Gateways
+
+As an example, 'eShopOnContainers' has around six internal microservice-types that have to be published through the API Gateways, as shown in the following image.
+ 
+![](./media/image29.png)
+
+**Figure 8-28.** Microservice folders in eShopOnContainers solution in Visual Studio
+
+About the Identity service, in the design it is left out of the API Gateway routing because it is the only cross-cutting concern in the system, but with Ocelot it is also possible to include it as part of the rerouting lists.
+
+All those services are currently implemented as ASP.NET Core Web API services, as you can tell because of the code. Let’s focus on one of the microservices like the Catalog microservice code.
+
+![](./media/image30.png)
+
+**Figure 8-29.** Sample Web API microservice (Catalog microservice)
+
+You can see that the Catalog microservice is a typical ASP.NET Core Web API project with several controllers and methods like in the following code.
+
+```csharp
+[HttpGet]
+[Route("items/{id:int}")]
+[ProducesResponseType((int)HttpStatusCode.NotFound)]
+[ProducesResponseType(typeof(CatalogItem),(int)HttpStatusCode.OK)]
+public async Task<IActionResult> GetItemById(int id)
+{
+    if (id <= 0)
+    {
+        return BadRequest();
+    }
+    var item = await _catalogContext.CatalogItems.
+                                          SingleOrDefaultAsync(ci => ci.Id == id);
+    //…
+
+    if (item != null)
+    {
+        return Ok(item);
+    }
+    return NotFound();
+}
+```
+The Http request will end up running that kind of C# code accessing the microservice database plus any additional action.
```

Comment:
add blank line after the code

Http -> HTTP

Reply:
👍